### ISSUE TITLE
Is computation and communication overlapped in PyTorch backend?

### ISSUE THREAD

**Issue by stgzr** [horovod/horovod#665]
Hi,
I wonder if there is a way to overlap computation and communication in PyTorch, that could be very useful in distributed training.
Thank you!

**Comment by tgaddair** [comment]
Hey @stgzr, there is some amount of overlap because every request to allreduce a gradient is made asynchronously in Horovod.  This allows us to batch tensors together to be allreduced all at once, what we call _tensor fusion_.  

Not sure if that answers your question, feel free to add more detail if not!

**Comment by shyhuai** [comment]
Hi, it seems that the [hooks](https://github.com/uber/horovod/blob/master/horovod/torch/__init__.py#L113) can only be started after all the gradients from the last layer to the first layer have been calculated. I added some debug information with timestamp in the hook function, and the printed timestamps in each iteration are nearly at the same value.  I am not sure if I did something wrong. If not, even allreduction of a gradient is made asynchronously, gradient computation and communication is not overlapped.

**Comment by alsrgv** [comment]
@shyhuai, if you're willing to invest time into investigating this further, could you run nvprof and see if there's overlap between computation and NCCL at CUDA kernel level?

**Comment by JF-D** [comment]
I wonder if this question is solved? I test the distributed performance for pytorch and tensorflow, it seems that computation and communication are overlapped. But I am not sure whether they are executed in different streams.

**Comment by tgaddair** [comment]
Hey @JF-D, we do allocate a separate CUDA stream (optionally multiple streams) in Horovod separate from the one used by PyTorch.

**Comment by sj6077** [comment]
@tgaddair What about TensorFlow? Is it also valid to TensorFlow?
